### PR TITLE
[Arctic-188][Flink] Avoiding submitting too much empty snapshots

### PR DESCRIPTION
**Pull request by YesOrNo828**
Issue: #188 

**Main change:**
Flink1.12 and flink1.14 **ArcticFileWriter** judge whether the WriteResult is empty or not.

### CODE REVIEW

**Comment by SteNicholas** [comment]
@YesOrNo828, could you please add some tests to verify whether not to submit empty snapshots? BTW, I was confused about how to produce this situation for empty results?

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 1]
```diff
@@ -199,7 +198,17 @@ public void close() throws Exception {
...
-    output.collect(new StreamRecord<>(writeResult));
+    if (!isWriteResultEmpty(writeResult)) {
+      // Only emit a non-empty WriteResult to committer operator, thus avoiding submitting too much empty snapshots.
+      output.collect(new StreamRecord<>(writeResult));
+    }
+  }
+
+  private boolean isWriteResultEmpty(WriteResult writeResult) {
+    return writeResult == null ||
+        (writeResult.dataFiles().length == 0 &&
```
Comment (zstraw) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:209]: I recommend org.apache.commons.lang.ArrayUtils.isEmpty to check not null

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 2]
```diff
@@ -198,7 +198,17 @@ public void dispose() throws Exception {
...
 
   private void emit(WriteResult writeResult) {
-    output.collect(new StreamRecord<>(writeResult));
+    if (!isWriteResultEmpty(writeResult)) {
+      // Only emit a non-empty WriteResult to committer operator, thus avoiding submitting too much empty snapshots.
+      output.collect(new StreamRecord<>(writeResult));
+    }
+  }
+
+  private boolean isWriteResultEmpty(WriteResult writeResult) {
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:207]: It's recommended to give a method directly named `isWriteResultNotEmpty`.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 3]
```diff
@@ -198,7 +198,17 @@ public void dispose() throws Exception {
   }
 
   private void emit(WriteResult writeResult) {
-    output.collect(new StreamRecord<>(writeResult));
+    if (!isWriteResultEmpty(writeResult)) {
+      // Only emit a non-empty WriteResult to committer operator, thus avoiding submitting too much empty snapshots.
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:202]: It's better to give a comment about the situation hoe the empty `WriteResult` produces.

**Comment by YesOrNo828** [comment]
@SteNicholas  @zstraw 
 I added related tests and a configuration to control whether to emit the empty WriteResults.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 4]
```diff
@@ -199,7 +205,27 @@ public void close() throws Exception {
...
+
+  /**
+   * Whether to emit the WriteResult.
+   *
+   * @param writeResult the WriteResult to emit
+   * @return true if the WriteResult should be emitted,
+   *         false only if the WriteResult is empty and the submitEmptySnapshot is false.
+   */
+  private boolean shouldEmit(WriteResult writeResult) {
+    if (submitEmptySnapshot) {
```
Comment (SteNicholas) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:222]:
```
return submitEmptySnapshot || (writeResult != null &&
        (!ArrayUtils.isEmpty(writeResult.dataFiles()) ||
            !ArrayUtils.isEmpty(writeResult.deleteFiles()) ||
            !ArrayUtils.isEmpty(writeResult.referencedDataFiles())));

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 5]
```diff
@@ -199,7 +205,27 @@ public void close() throws Exception {
...
+      // Only emit a non-empty WriteResult to committer operator, thus avoiding submitting too much empty snapshots.
+      output.collect(new StreamRecord<>(writeResult));
+    }
+  }
+
+  /**
+   * Whether to emit the WriteResult.
+   *
+   * @param writeResult the WriteResult to emit
+   * @return true if the WriteResult should be emitted,
```
Comment (SteNicholas) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:218]:
```suggestion
   * @return true if the WriteResult should be emitted or the WriteResult isn't empty,
```

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 6]
```diff
@@ -198,7 +204,27 @@ public void dispose() throws Exception {
...
+      output.collect(new StreamRecord<>(writeResult));
+    }
+  }
+
+  /**
+   * Whether to emit the WriteResult.
+   *
+   * @param writeResult the WriteResult to emit
+   * @return true if the WriteResult should be emitted,
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:217]:
```suggestion
   * @return true if the WriteResult should be emitted or the WriteResult isn't empty,
```

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 7]
```diff
@@ -198,7 +204,27 @@ public void dispose() throws Exception {
...
+   *         false only if the WriteResult is empty and the submitEmptySnapshot is false.
+   */
+  private boolean shouldEmit(WriteResult writeResult) {
+    if (submitEmptySnapshot) {
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:221]:
```
return submitEmptySnapshot || (writeResult != null &&
        (!ArrayUtils.isEmpty(writeResult.dataFiles()) ||
            !ArrayUtils.isEmpty(writeResult.deleteFiles()) ||
            !ArrayUtils.isEmpty(writeResult.referencedDataFiles())));
```

**Review by zstraw — APPROVED** [review]

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**